File: packages/agentic/desktop-driver/native/windows/src/inputlock.cpp
#include "inputlock.h"

#include <iostream>
#include <mutex>
#include <string>
#include <utility>

namespace furn {
namespace {

// The named mutex handle lives for the life of the process; ownership is
// tracked per thread so nested command scopes never wait twice.
std::mutex stateMutex;
HANDLE inputMutex = nullptr;
DWORD ownerThread = 0;
unsigned int ownerDepth = 0;
std::wstring inputMutexName = kPhysicalInputMutexName;

HANDLE EnsureMutex() {
  if (inputMutex == nullptr) {
    inputMutex = CreateMutexW(nullptr, FALSE, inputMutexName.c_str());
    if (inputMutex == nullptr) {
      FailLastError(kErrorInputBusy, "Creating the physical input mutex", GetLastError());
    }
  }
  return inputMutex;
}

void WriteDiagnostic(const std::string& message) {
  std::cerr << "furn-desktop-driver-host: " << message << '\n';
}

}  // namespace

const wchar_t* PhysicalInputMutexName() noexcept {
  return inputMutexName.c_str();
}

void SetPhysicalInputMutexNameForTesting(std::wstring name) {
  const std::lock_guard<std::mutex> guard(stateMutex);
  if (name.empty() || ownerDepth != 0) {
    Fail(kErrorInvalidParams, "The physical input mutex test name is invalid or the mutex is currently owned.");
  }
  if (inputMutex != nullptr) {
    CloseHandle(inputMutex);
    inputMutex = nullptr;
  }
  inputMutexName = std::move(name);
  ownerThread = 0;
}

PhysicalInputScope::PhysicalInputScope(const CancellationToken& token, unsigned int timeoutMs, AbandonPolicy policy) {
  HANDLE handle = nullptr;
  {
    const std::lock_guard<std::mutex> guard(stateMutex);
    handle = EnsureMutex();
    if (ownerDepth > 0 && ownerThread == GetCurrentThreadId()) {
      ownerDepth += 1;
      owns_ = true;
      nested_ = true;
      return;
    }
  }

  const Deadline deadline(timeoutMs);
  while (true) {
    token.ThrowIfCancelled();
    const DWORD result = WaitForSingleObject(handle, 50);
    if (result == WAIT_OBJECT_0) {
      break;
    }
    if (result == WAIT_ABANDONED) {
      if (policy == AbandonPolicy::Adopt) {
        WriteDiagnostic("adopting physical input abandoned by a terminated helper.");
        break;
      }
      ReleaseMutex(handle);
      Fail(kErrorInputBusy,
           "A previous desktop driver helper terminated while holding physical input, so keys or buttons may still "
           "be depressed. Run the helper with --release-input to recover, then retry the command.");
    }
    if (result != WAIT_TIMEOUT) {
      FailLastError(kErrorInputBusy, "Waiting for the physical input mutex", GetLastError());
    }
    if (deadline.Expired()) {
      Fail(kErrorInputBusy, "Another desktop driver helper held physical input for longer than " +
                                std::to_string(timeoutMs) +
                                " ms. Wait for the other session to finish its input command, or run the helper with "
                                "--release-input if that session is gone.");
    }
  }

  const std::lock_guard<std::mutex> guard(stateMutex);
  ownerThread = GetCurrentThreadId();
  ownerDepth = 1;
  owns_ = true;
}

PhysicalInputScope::PhysicalInputScope(PhysicalInputScope&& other) noexcept
    : owns_(other.owns_), nested_(other.nested_) {
  other.owns_ = false;
  other.nested_ = false;
}

PhysicalInputScope& PhysicalInputScope::operator=(PhysicalInputScope&& other) noexcept {
  if (this != &other) {
    Release();
    owns_ = other.owns_;
    nested_ = other.nested_;
    other.owns_ = false;
    other.nested_ = false;
  }
  return *this;
}

PhysicalInputScope::~PhysicalInputScope() {
  Release();
}

void PhysicalInputScope::Release() noexcept {
  if (!owns_) {
    return;
  }
  owns_ = false;
  nested_ = false;
  const std::lock_guard<std::mutex> guard(stateMutex);
  if (ownerDepth == 0 || ownerThread != GetCurrentThreadId()) {
    return;
  }
  ownerDepth -= 1;
  if (ownerDepth == 0) {
    ownerThread = 0;
    ReleaseMutex(inputMutex);
  }
}

}  // namespace furn

File: packages/agentic/desktop-driver/native/macos/Sources/DesktopDriverHost/InputLock.swift
import Foundation
import Darwin

enum AbandonPolicy {
  case fail
  case adopt
}

struct InputLockOwner: Codable {
  let pid: Int32
  let processStartedAt: String
}

final class PhysicalInputLock: @unchecked Sendable {
  static let shared = PhysicalInputLock()

  private let stateLock = NSLock()
  private var depth = 0
  private var descriptor: Int32 = -1
  private let lockURL: URL
  private let ownerURL: URL

  private enum OwnerState {
    case missing
    case valid(InputLockOwner)
    case invalid
  }

  init(baseName: String = "furn-desktop-driver-physical-input-v1-\(getuid())") {
    let directory = FileManager.default.temporaryDirectory
    lockURL = directory.appendingPathComponent(baseName + ".lock", isDirectory: false)
    ownerURL = directory.appendingPathComponent(baseName + ".owner.json", isDirectory: false)
  }

  func withLock<T>(
    token: CancellationToken,
    timeoutMilliseconds: Int = 30_000,
    policy: AbandonPolicy = .fail,
    _ body: () throws -> T
  ) throws -> T {
    try acquire(token: token, timeoutMilliseconds: timeoutMilliseconds, policy: policy)
    defer { release() }
    return try body()
  }

  func acquire(token: CancellationToken, timeoutMilliseconds: Int, policy: AbandonPolicy) throws {
    stateLock.lock()
    if depth > 0 {
      depth += 1
      stateLock.unlock()
      return
    }
    stateLock.unlock()

    let fd = open(lockURL.path, O_CREAT | O_RDWR | O_CLOEXEC, S_IRUSR | S_IWUSR)
    guard fd >= 0 else {
      try fail(ErrorCode.inputBusy, "Creating the physical input lock failed with errno \(errno).")
    }
    let deadline = DispatchTime.now() + .milliseconds(timeoutMilliseconds)
    do {
      while flock(fd, LOCK_EX | LOCK_NB) != 0 {
        if errno != EWOULDBLOCK && errno != EAGAIN {
          try fail(ErrorCode.inputBusy, "Acquiring the physical input lock failed with errno \(errno).")
        }
        try token.throwIfCancelled()
        if DispatchTime.now() >= deadline {
          try fail(
            ErrorCode.inputBusy,
            "Another desktop driver helper held physical input for longer than \(timeoutMilliseconds) ms. Wait for that session to finish, or run the helper with --release-input if it is gone."
          )
        }
        try token.wait(milliseconds: 50)
      }

      switch readOwner() {
      case let .valid(owner):
        let ownerStart = parseDate(owner.processStartedAt)
        let actualStart = processStartTime(owner.pid)
        let sameIdentity = ownerStart != nil && actualStart != nil
          && abs(ownerStart!.timeIntervalSince(actualStart!)) <= 0.01
        if processIsAlive(owner.pid) && sameIdentity {
          try fail(
            ErrorCode.inputBusy,
            "The physical input owner record still names a live desktop driver helper. Wait for it to finish before retrying."
          )
        }
        if policy == .fail {
          try fail(
            ErrorCode.inputBusy,
            "A previous desktop driver helper terminated while holding physical input, so keys or buttons may still be depressed. Run the helper with --release-input to recover, then retry the command."
          )
        }
        writeDiagnostic("furn-desktop-driver-host: adopting physical input abandoned by a terminated helper.")
      case .invalid:
        if policy == .fail {
          try fail(
            ErrorCode.inputBusy,
            "The physical input owner record is invalid, so prior input cleanup cannot be proven. Run the helper with --release-input to recover, then retry the command."
          )
        }
        writeDiagnostic("furn-desktop-driver-host: adopting an invalid physical input owner record for recovery.")
      case .missing:
        break
      }

      try writeOwner()
      stateLock.lock()
      descriptor = fd
      depth = 1
      stateLock.unlock()
    } catch {
      _ = flock(fd, LOCK_UN)
      close(fd)
      throw error
    }
  }

  func release() {
    stateLock.lock()
    guard depth > 0 else {
      stateLock.unlock()
      return
    }
    depth -= 1
    if depth > 0 {
      stateLock.unlock()
      return
    }
    let fd = descriptor
    descriptor = -1
    stateLock.unlock()

    try? FileManager.default.removeItem(at: ownerURL)
    _ = flock(fd, LOCK_UN)
    close(fd)
  }

  private func readOwner() -> OwnerState {
    guard FileManager.default.fileExists(atPath: ownerURL.path) else {
      return .missing
    }
    guard let data = try? Data(contentsOf: ownerURL), !data.isEmpty, data.count <= 4096 else {
      return .invalid
    }
    guard let owner = try? JSONDecoder().decode(InputLockOwner.self, from: data) else {
      return .invalid
    }
    return .valid(owner)
  }

  private func writeOwner() throws {
    guard let start = processStartTime(getpid()) else {
      try fail(ErrorCode.inputBusy, "The helper could not read its process start time for physical input ownership.")
    }
    let owner = InputLockOwner(pid: getpid(), processStartedAt: formatDate(start))
    do {
      let data = try JSONEncoder().encode(owner)
      try data.write(to: ownerURL, options: [.atomic])
      try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: ownerURL.path)
    } catch {
      try fail(ErrorCode.inputBusy, "Writing the physical input owner record failed: \(error.localizedDescription)")
    }
  }

  func installDeadOwnerForSelfTest() throws {
    let owner = InputLockOwner(pid: Int32.max, processStartedAt: "1970-01-01T00:00:00.000Z")
    try JSONEncoder().encode(owner).write(to: ownerURL, options: [.atomic])
  }

  func cleanupSelfTestFiles() {
    try? FileManager.default.removeItem(at: ownerURL)
    try? FileManager.default.removeItem(at: lockURL)
  }
}
